[InstCombine] Fold (X==Z) ? (Y==Z) : (!(Y==Z) && X==Y) --> X==Y

citymarina opened the pull request:
This corresponds to the canonicalized form of some logic that was
seen in Swift-generated code for comparing optional pointers:
`(X==Z || Y==Z) ? (X==Z && Y==Z) : X==Y --> X==Y`
where `Z` was the constant `0`.

https://alive2.llvm.org/ce/z/J_3aa9

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1406,6 +1406,46 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+  Value *Either = Sel.getCondition(), *Both = Sel.getTrueValue(),
+        *Cmp = Sel.getFalseValue();
+
+  if (!match(Either, m_Select(m_Value(XEq), m_One(), m_Value(YEq))))
```
`m_LogicalOr(m_Value(XEq), m_Value(YEq)`

citymarina replied:
Thanks, I didn't know about these.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1406,6 +1406,46 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+    return nullptr;
+
+  if (!match(Both, m_Select(m_Specific(XEq), m_Specific(YEq), m_Zero())) &&
+      !match(Both, m_Select(m_Specific(YEq), m_Specific(XEq), m_Zero())))
```
`m_LogicalAnd(m_Specific(XEq), m_Specific(YEq)`

citymarina replied:
I needed to use the commutative version here; I assume that's ok.

goldsteinn commented on llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp:
```diff
@@ -1406,6 +1406,46 @@ Instruction *InstCombinerImpl::foldSelectValueEquivalence(SelectInst &Sel,
+    return nullptr;
+
+  if (!match(XEq,
+             m_c_SpecificICmp(ICmpInst::ICMP_EQ, m_Value(X), m_Value(Z))) ||
```
The commutative here isn't doing anything since your just match any value on either side.

goldsteinn replied:
This should look something like:
```
if(!match(XEq, m_SpecificICmp(ICMP_EQ, m_Value(X), m_Value(Z))
  return nullptr;
if(!match(m_c_SpecificICmp(ICMP_EQ, m_Value(Y), m_Specific(Z)))
  std::swap(X, Z);
if(!match(m_c_SpecificICmp(ICMP_EQ, m_Value(Y), m_Specific(Z)))
  return nullptr;
```

citymarina replied:
Ah, this points out a case that my tests were missing: `icmp eq 0, %Y` got canonicalized to have the zero on the RHS, so I wasn't covering the case that needed `swap`. I've added a case that uses X, Y, Z to avoid this.

goldsteinn commented:
This is really really specific, if it shows up in realworld code I'm not opposed to getting it in, but is there some generalization of the pattern we can make?

citymarina commented:
> This is really really specific, if it shows up in realworld code I'm not opposed to getting it in, but is there some generalization of the pattern we can make?

Oh, interesting. I was worried that there would be concern about having overgeneralized already (my original unpublished version just checked for constant 0, with no commutativity, etc.). I've tried to relax as many of the constraints as I could see; do you have anything specific in mind about how to take it further?

goldsteinn commented:
> > This is really really specific, if it shows up in realworld code I'm not opposed to getting it in, but is there some generalization of the pattern we can make?
> 
> Oh, interesting. I was worried that there would be concern about having overgeneralized already (my original unpublished version just checked for constant 0, with no commutativity, etc.). I've tried to relax as many of the constraints as I could see; do you have anything specific in mind about how to take it further?

I don't really see any obvious generalization either. If its backed up by a realworld use-case its probably fine.

nikic reviewed (commented):
It looks like InstCombine canonicalizes this to something like `(X == Z) ? (Y == Z) : !(Y == Z) && X == Y)`, see https://alive2.llvm.org/ce/z/XY_HMX. Shouldn't we be matching that pattern instead?

citymarina commented:
> It looks like InstCombine canonicalizes this to something like `(X == Z) ? (Y == Z) : !(Y == Z) && X == Y)`, see https://alive2.llvm.org/ce/z/XY_HMX. Shouldn't we be matching that pattern instead?

I am new in this area and unfamiliar with the conventions. My thought process was that matching something closer to the Swift-generated IR would better express the intent of this patch, and also I wasn't sure if I could rely on the altered form remaining stable over time. But if you think it would be better then I'm happy to update the patch.

goldsteinn commented:
> > It looks like InstCombine canonicalizes this to something like `(X == Z) ? (Y == Z) : !(Y == Z) && X == Y)`, see https://alive2.llvm.org/ce/z/XY_HMX. Shouldn't we be matching that pattern instead?
> 
> I am new in this area and unfamiliar with the conventions. My thought process was that matching something closer to the Swift-generated IR would better express the intent of this patch, and also I wasn't sure if I could rely on the altered form remaining stable over time. But if you think it would be better then I'm happy to update the patch.

Typically canonical patterns are what remain stable longer. They aren't folds that meaningful change the instructions, just standardize them in a way so that future matching code only needs to worry about a single iteration.
You can see more about it: https://llvm.org/docs/InstCombineContributorGuide.html#canonicalization-and-target-independence

citymarina commented:
> > > It looks like InstCombine canonicalizes this to something like `(X == Z) ? (Y == Z) : !(Y == Z) && X == Y)`, see https://alive2.llvm.org/ce/z/XY_HMX. Shouldn't we be matching that pattern instead?
> > 
> > 
> > I am new in this area and unfamiliar with the conventions. My thought process was that matching something closer to the Swift-generated IR would better express the intent of this patch, and also I wasn't sure if I could rely on the altered form remaining stable over time. But if you think it would be better then I'm happy to update the patch.
> 
> Typically canonical patterns are what remain stable longer. They aren't folds that meaningful change the instructions, just standardize them in a way so that future matching code only needs to worry about a single iteration. You can see more about it: https://llvm.org/docs/InstCombineContributorGuide.html#canonicalization-and-target-independence

That makes sense. I guess the missing piece for me was that I didn't realize `(X == Z) ? (Y == Z) : !(Y == Z) && X == Y)` was in fact a canonicalization. I'll work on an update.

citymarina committed:
[InstCombine] Add tests for folding equality test. NFC

citymarina committed:
[InstCombine] Fold `(X==Z) ? (Y==Z) : (!(Y==Z) && X==Y) --> X==Y`

This corresponds to the canonicalized form of some logic that was
seen in Swift-generated code for comparing optional pointers:
`(X==Z || Y==Z) ? (X==Z && Y==Z) : X==Y --> X==Y`
where `Z` was the constant `0`.

https://alive2.llvm.org/ce/z/J_3aa9

citymarina commented:
Ping? Thanks.

goldsteinn commented:
This could (but doesn't) work for cases where the compare ops are inverted for example:
https://alive2.llvm.org/ce/z/6BD9mt

Since this is motivated by a pretty specific case where we always get equal expressions I'm not sure this is an issue worth fixing given the extra code complexity required to do so.

goldsteinn reviewed (approved):
LGTM. Wait a day or so to push to give others a chance to take a look.

nikic reviewed (commented):
Could you please update the alive2 proof in the PR description? Otherwise this looks good to me.

fhahn commented:
Looks like the alive2 proof in the description has been updated, thanks!
